services: Redesign details page

marusak opened the pull request:
Redesign and rewrite (part of) service details into react. (Only the panel marked with red is in React as I wanted to make as little changes as possible, moving other things into React should be follow ups).
Follows [this](https://raw.githubusercontent.com/cockpit-project/cockpit-design/master/services/service-details.png) design.

No need for code review, I plan to polish that anyway. Firstly I would like to get design review by @garrett as he created the design. Can you please take a look, thanks. If you have some small styling fixes feel free to push them here (ideally as separate commit(s)). Also @andreasn if you have time I would like to hear your feedback as well, it is different to use it than to see mock-up and you can maybe find some rough edges that could be fixed. Thanks. 

Only functional change that is missing and I know of is not showing `vendor` related stuff. Still havn't found reasonable way of getting this info.

![screenservices](https://user-images.githubusercontent.com/12330670/60886024-2ea31080-a251-11e9-8540-7373a63ee927.png)

TODO:

- [ ] Implement `vendor` stuff
- [x] Use ct-form-layout instead of form-table-ct
- [x] Tooltip for on-off switch
- [x] Accessibility review
   - [x] Keyboard arrows are unpredictable (pressing down skipped the second menu item)
- [x] When kebab menu is opened the dropdown does not have arrow. #12290
- [x] Write proper message when masking service.
- [x] Previously in the list of relationships unit possibly could be disabled if does not exist. Is this possible? (runits)
- [x] Tests
- Browser compatibility
   - [x] Firefox
   - [x] Chrome
   - [x] Epiphany (WebKit)
   - [x] Edge
- [x] Fix delay when clicking

andreasn commented:
What's _vendor_ exactly?
I don't see that in the mockup

andreasn commented on pkg/systemd/service-details.css:
```diff
@@ -0,0 +1,63 @@
+}
+
+.green-icon {
+    color: var(--pf-global--success-color--100);
```
This color creates very little contrast against the white, so I wonder if `--pf-global--success-color--200` would be a better choice here.

andreasn replied:
But that's more of a gut feeling on my side. Not sure how to check it against the [contrast ratios mentioned here](https://www.patternfly.org/v4/design-guidelines/styles/colors)

marusak replied:
Yeah, the colors and font sizes I picked were mostly random and I am open to change those :) I think this one is the only one that I picked because it is default green color for patternfly icons, but agree, I also think this is not green enough.

andreasn replied:
Oh, interesting. Do you know where the guidelines for that is?
I was looking at this page, under Alert statuses > Success > Colors in Context

marusak replied:
I don't really know, I just picked it up when I imported `pficon-ok` from https://www.patternfly.org/styles/icons/ and with inspector looked at its color.

marusak replied:
I changed it to `--pf-global--success-color--200` but it does not seem that distinguishable which are green and which are grey.
![notso2](https://user-images.githubusercontent.com/12330670/60955878-ff98a780-a301-11e9-88f1-d50c414d6a10.png)

marusak replied:
But yeah, is less distractive, especially with red one:
![withred](https://user-images.githubusercontent.com/12330670/60956112-72098780-a302-11e9-8941-80eaeb9d0fc4.png)

andreasn commented on pkg/systemd/service-details.jsx:
```diff
@@ -0,0 +1,445 @@
+        this.onOnOffSwitch = this.onOnOffSwitch.bind(this);
+    }
+
+    onOnOffSwitch() {
```
With this, we're intentionally moving to a control that doesn't express it's full intent as much as the old dropdown did, for good and bad. I think it could help with a tooltip of what's it's doing. Like _Start and Enable_ or something like that.

marusak replied:
Done, there is tooltip.

garrett commented:
This is a great start.

- I've opened a PR to rename `ct-form-layout` to `ct-form` (#11966), so if that lands before this PR, you'll want to update. (It's super-simple — it involves dropping the `-layout` part from the name.)
- As mentioned in IRC, Composer has the ^ part of the popup. I'm not sure what's different.
- Various style changes
  - The top part is in a box. It probably shouldn't be. (I didn't have it in my mockup in a box and other parts of Cockpit don't have everything in boxes.)
  - There are no spaces between the commas and the next item. (This appears to be CSS related and I'm happy to fix this up if desired.)

What's your intended scope for this PR? Is it going to stay the highlighted part only? Will it be all of the details page? Is it going to include the list on the parent page? Or will it be broken into additional PRs?

Again, just like I said on IRC: I'm really happy to see this being worked on. Hopefully this will be a lot clearer for our users. Thanks for implementing it! :+1:

andreasn commented:
This is how it would look without the white box behind it:
![Screenshot from 2019-07-09 16-11-58](https://user-images.githubusercontent.com/209895/60895111-5780d100-a264-11e9-93f0-b183a47e7354.png)

marusak commented:
> The top part is in a box. It probably shouldn't be. (I didn't have it in my mockup in a box and other parts of Cockpit don't have everything in boxes.)

Yeah, this is one think where I moved away from the design. On my low-contrast monitor the greyish background and black letter (and grey icons) did not have enough contrast. I'll get more contrast and try again :)

> What's your intended scope for this PR? Is it going to stay the highlighted part only? 

This is it. Nothing more in this PR, it is kinda big anyway (and no tests yet).  Only implementing/fixing the TODO list items and polishing css. When this lands I am more than happy to continue with other parts of services :)

garrett commented:
> did not have enough contrast

The contrast far surpasses AAA, with a contrast ratio of 17.5 (and required for AAA is 7... AA is just 4.5 or higher).

From https://accessible-colors.com/:

![Screenshot_2019-07-09 Accessible Colors WCAG 2 0 AA and AAA color contrast checker](https://user-images.githubusercontent.com/10246/60898790-b137ca00-a269-11e9-8b42-210c87354377.png)

> This is it. […] When this lands I am more than happy to continue […]

Nice. It's good to do thing like this in pieces.

(I'm sure @andreasn is now laughing at me saying this, considering all my recent touch-all-the-files-of-Cockpit with literally _hundreds_ of commits (before squashing down to _several_). :laughing:)

I'm looking forward to this landing soon and the rest of the updates. Thanks again!

garrett commented:
> moved away from the design

Well, I kinda messed up all the designs with the PF4-like updates. :wink: So it's all open to interpretation now.

marusak commented:
![startingservice](https://user-images.githubusercontent.com/12330670/61111086-09004c00-a489-11e9-94c4-31ce11bd4153.png)
 I think the design has one issue - when service is enabled but stopped there is no way how to start it. (toggling the switch off and back on is not a right way of doing it, also restart is confusing). When service failed, there is button to start it. I am thinking if this button should be explicit for stopped service as well or just add it into the kebab menu. I think that replacing "restart" with "start" in the kebab menu would be right, so I'll do that. @garrett @andreasn  what do you think?

andreasn commented:
>  I think that replacing "restart" with "start" in the kebab menu would be right, so I'll do that.

I agree, it's more contextually correct in that case, since you can't restart something that isn't started.
Let's try keeping it in the kebab for now. If it turns out in usability testing it's too hidden, we'll move it to a more prominent place.

andreasn commented:
>  I think that replacing "restart" with "start" in the kebab menu would be right, so I'll do that.

I agree, it's more contextually correct in that case, since you can't restart something that isn't started.
Let's try keeping it in the kebab for now. If it turns out in usability testing it's too hidden, we'll move it to a more prominent place.

marusak commented:
I've fixed a few things as well as those mentioned here in comments. Also wrote tests and now running a first run to see what needs to be addressed.

Could I get design and accessibility review by @garrett ? 
Also could I get code review from @KKoukiou or @martinpitt?
Thank you

Updated screenshot:
![newstyle](https://user-images.githubusercontent.com/12330670/61200890-2d576500-a6e3-11e9-9d75-830827c2468c.png)

garrett commented:
> when service is enabled but stopped there is no way how to start it

Yeah, I agree about the kebab menu having "start" instead of restart. (It was an oversight on my part during the design — there were a lot of iterations to step through. Thanks for fixing that!) :+1:

If it's not obvious enough, we could _also_ have a "start service" button similar to the "failed to start" status. But, the reason why it's not there is because in many cases, you might have something enabled and run once on startup, then it exits afterward. In those cases, you don't want to coax the user to re-run the service if there's nothing wrong. So, basically... I think we should still leave out the button when there's nothing wrong happening.

garrett commented:
The last comment made me wonder if it makes sense to add a "Last ran on ..." text to the enabled & stopped status. It would probably be even more interesting to know it ran on the current boot (at boot time), perhaps. :man_shrugging: It's probably fine without it still though. :wink:

garrett commented:
When I view service details, go back to the list, and view details from another service, I see the previous service for a second until the new one pops in.

marusak commented on pkg/systemd/service-details.jsx:
```diff
@@ -0,0 +1,436 @@
+                                          confirmText={ _("Mask Service") }
+                                          confirmAction={() => { this.props.fileActionCallback("MaskUnitFiles", false); this.setState({ dialogMaskedOpened: false }) }} />
+                }
+                <DropdownKebab id="service-actions" title={ _("Additions actions") }>
```
echm, spell this correctly on next push :D

marusak commented:
> When I view service details, go back to the list, and view details from another service, I see the previous service for a second until the new one pops in.

Fixed. Maybe it could do with some curtains showing spinner. Possible follow up?

marusak commented:
Shows spinner and disabled actions while any action (created from cockpit) in progress:
![Screenshot from 2019-07-16 14-46-38](https://user-images.githubusercontent.com/12330670/61295621-c1efbf00-a7d8-11e9-9925-5c4172177441.png)

garrett commented:
It's working better! This is great.

I think we want to remove "running" and "automatically starts" (and anything like this) from the status area while it has a spinner.

I noticed it:
1. added a spinner to the status area (and didn't remoove the status)
2. swapped out the status with the new status (and kept the spinner)
3. removed the spinner

Instead, it should:
1. replace the status with the spinner
2. replace the spinner with the new status

Also, there should be a space between the spinner and the updating text.

marusak commented:
> Instead, it should: ...

Done :)

I also fixed the last issue about the select jumping two rows instead of one. The only remaining thing is implementing vendor related stuff, but I am keeping that one as a follow-up (it is not in current implementation either and I still was not properly able to get this information from dbus API).

This is ready for code review.

KKoukiou commented:
@marusak is this related here? https://209.132.184.41:8493/logs/pull-12265-20190716-192929-7550a36e-verify-fedora-atomic/log.html#234

KKoukiou commented:
> > is this related here?
> 
> hmm, I don't think so. I've run it a few time locally and could not reproduce. Also from the logic of the test this should not happen (you enable and stop unit, check in overview it is so and then check it with details it is also like that - which uses the same values to display both overview and details). I've retried it here as well, let's see if it can hit it again or it was one off.

On retry the test didn't fail again. But now there is some conflict, can you please rebase?

marusak commented:
> But now there is some conflict, can you please rebase?

Ah forgot to do it when the other PR for services landed. Fixed.

KKoukiou commented on pkg/systemd/service-details.jsx:
```diff
@@ -0,0 +1,506 @@
+ *      Method for calling unit methods like `UnitStart`
+ *  - fileActionCallback
+ *      Method for calling unit file methods like `EnableUnitFiles`
+ *  - disbaled
```
typo here

KKoukiou commented on pkg/systemd/service-details.jsx:
```diff
@@ -0,0 +1,506 @@
+                    actions.push(
+                        <MenuItem key="divider1" divider />
+                    );
+                if (this.props.active)
```
nitpick; We generally tend to put the if else bodys inside curly braces when they are multiline.

marusak commented:
Thank you @KKoukiou for the review. Fixed and pushed again.

marusak commented:
Newest screenshot:
![screenshotservices](https://user-images.githubusercontent.com/12330670/61534953-46d51580-aa31-11e9-9f22-41168a6ebcac.png)
